ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java

```diff
@@ -189,6 +191,30 @@ private void analyzeNonStdOutSink(final Sink sink) {
           .setInto(Into.newSink(sink.getName(), topicName, windowInfo, keyFmtInfo, valueFmtInfo));
     }
 
+    private FormatInfo buildKeyFormatInfo(
+        final Optional<String> explicitFormat,
+        final Map<String, String> formatProperties,
+        final FormatInfo sourceFormat
+    ) {
+      final boolean partitioningByNull = analysis.getPartitionBy()
+          .map(pb -> pb.getExpression() instanceof NullLiteral)
+          .orElse(false);
+
+      if (partitioningByNull) {
+        final boolean nonNoneExplicitFormat = explicitFormat
+            .map(fmt -> !fmt.equalsIgnoreCase(NoneFormat.NAME))
+            .orElse(false);
+
+        if (nonNoneExplicitFormat) {
+          throw new KsqlException("Key format specified for stream without key columns.");
```

Comment:
why is this an error? can't I have `null` keys with non-`NONE` formats?

Reply:
If the key format supports schema inference then it's going to blow up later when trying to build the schema for the key.